### ISSUE TITLE
ames: +dedup-message takes a long time on busy ships

### ISSUE THREAD

**Issue by philipcmonk** [urbit/urbit#6308]
While investigating another issue, we realized that on ~mister-dister-dozzod-dozzod, +dedup-message takes around 800ms.  For reference, packets normally take around 2ms, and any time clay sends a file larger than 1KB, +dedup-message will be called.  This means that a very large portion of the OTA time is spent in +dedup-message.

+dedup-message takes a message that we intend to send and scans through all our other outstanding messages to find a duplicate of it, and if it finds it, it uses that one.  This reduces memory usage significantly, and we can't just remove it (although there may be more invasive replacements available).

Currently, +dedup-message is written as a linear search across all peers, all bones, all messages.  One way to make this much faster would be to maintain a map or set of outstanding messages keyed by content -- then lookup would be log(n), almost certainly fast enough.  Adding to this map is easy, but it would need to be garbage-collected or reference-counted.  Perhaps the easiest way would be to throw away and rebuild the map every few minutes -- this would take about as long as +dedup-message does now, so you would spend 800ms every few minutes instead of constantly.

**Comment by joemfb** [comment]
I wonder if we can solve this another way.

We could add a memoizing `+jam` jet, using a new, dedicated, capped HAMT in the road, and update `%ames` to store non-duplicated noun messages instead of needing-deduplication jammed atoms. `%ames` would just call `(cut 13 [fragment-num 1] (memoized-jam msg))` for every fragment (re-)send. This would preserve existing structural sharing, and give us the flexibility to trade time for space on ships with memory pressure (by configuring a smaller cache-size on the HAMT and/or trimming it heuristically).

**Comment by tacryt-socryp** [comment]
Do these repeated calls to `+dedup-message` happen in separate events? `+jim` is memoized already, so any duplicate messages sent in the same event should be "pre-deduplicated".

**Comment by philipcmonk** [comment]
Yeah, they're in separate events. Typical case is an OTA: as each ship
hears that there's a new version, it asks for each of the files, and those
are relatively large. They used to be far larger when the whole OTA was one
message.

**Comment by belisarius222** [comment]
Remote scry should address the OTA part of this problem, so we could punt. If we do address it, deleting the deduplication routine and just re-jamming messages sounds like less complexity than adding more state and a GC routine to Ames.

The downside of that approach would be the need for a special HAMT cache, but doesn't the ~+ in +jim already work this way? The road that runs the kernel should persist for the whole Arvo event, so if Gall sends two facts to Ames in the same event, the second call to +jim should hit the HAMT cache for that road. Right?

Re-jamming in separate Arvo events would not be memoized, but I imagine +jam is faster than the dedup routine in most cases, and it would need less machinery.

**Comment by philipcmonk** [comment]
They're usually not sent in the same event
